clippy_lints/src/try_err.rs

```diff
@@ -0,0 +1,120 @@
+    /// can fail to be easily chained. For example, `foo()?.bar()` or
+    /// `foo(bar()?)`. Because `Err(x)?` can't be used that way (it will
+    /// always return), it is more clear to write `return Err(x)`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// fn foo(fail: bool) -> Result<i32, String> {
+    ///     if fail {
+    ///       Err("failed")?;
+    ///     }
+    ///     Ok(0)
+    /// }
+    ///
+    /// // Good
+    /// fn foo(fail: bool) -> Result<i32, String> {
+    ///     if fail {
+    ///       return Err("failed".into());
+    ///     }
+    ///     Ok(0)
+    /// }
+    /// ```
+    pub TRY_ERR,
+    style,
+    "return errors explicitly rather than hiding them behind a `?`"
+}
+
+declare_lint_pass!(TryErr => [TRY_ERR]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TryErr {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        // Looks for a structure like this:
+        // match ::std::ops::Try::into_result(Err(5)) {
+        //     ::std::result::Result::Err(err) =>
+        //         #[allow(unreachable_code)]
+        //         return ::std::ops::Try::from_error(::std::convert::From::from(err)),
+        //     ::std::result::Result::Ok(val) =>
+        //         #[allow(unreachable_code)]
+        //         val,
+        // };
+        if_chain! {
+            if let ExprKind::Match(ref match_arg, _, MatchSource::TryDesugar) = expr.node;
+            if let ExprKind::Call(ref match_fun, ref try_args) = match_arg.node;
+            if let ExprKind::Path(ref match_fun_path) = match_fun.node;
+            if match_qpath(match_fun_path, &["std", "ops", "Try", "into_result"]);
+            if let Some(ref try_arg) = try_args.get(0);
+            if let ExprKind::Call(ref err_fun, ref err_args) = try_arg.node;
+            if let Some(ref err_arg) = err_args.get(0);
+            if let ExprKind::Path(ref err_fun_path) = err_fun.node;
+            if match_qpath(err_fun_path, &paths::RESULT_ERR);
+            if let Some(return_type) = find_err_return_type(cx, &expr.node);
+
+            then {
+                let err_type = cx.tables.expr_ty(err_arg);
+                let suggestion = if err_type == return_type {
+                    format!("return Err({})", snippet(cx, err_arg.span, "_"))
+                } else {
+                    format!("return Err({}.into())", snippet(cx, err_arg.span, "_"))
+                };
+
+                span_lint_and_then(
+                    cx,
+                    TRY_ERR,
+                    expr.span,
+                    &format!("confusing error return, consider using `{}`", suggestion),
+                    |db| {
+                        db.span_suggestion(
+                            expr.span,
+                            "try this",
+                            suggestion,
+                            Applicability::MaybeIncorrect
```

Comment:
You should be able to use the `utils::span_sugg_and_lint` function directly.

Also, do you have an example where this lint suggests something incorrect? If not, you can turn this to `MachineApplicable` and add `// run-rustfix` to the test file, like here:
https://github.com/rust-lang/rust-clippy/blob/da7aebc342b7dae444912b98dedbab22bf2a224c/tests/ui/println_empty_string.rs#L1

Reply:
Maybe it's contrived, but a number of the test cases don't compile after applying the fix. For example:
```
pub fn basic_test() -> Result<i32, i32> {
    let err: i32 = 1;
    Err(err)?;
    Ok(0)
}
```
becomes
```
pub fn basic_test() -> Result<i32, i32> {
    let err: i32 = 1;
    return Err(err);
    Ok(0)
}
```
and the compiler complains about the unreachable `Ok(0)`. I don't expect that to come up much in real life, but the replacement isn't 100% safe.

Reply:
But that's just a warning, so it still compiles (except you pass `-Dwarnings` of course). Also the `unreachable_code` warning would also apply to the unfixed code, it was just hidden with the weird `Err(_)?` syntax. So `MachineApplicable` should be ok here IMO. 

Is `// run-rustfix` tested with `-Dwarnings`? Otherwise this can be added too.